pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix

```diff
@@ -0,0 +1,96 @@
+{ lib
+, buildPgrxExtension
+, callPackage
+, cargo-pgrx
+, clang_16
+, fetchCrate
+, fetchFromGitHub
+, nix-update-script
+, openssl
+, pkg-config
+, postgresql
+, rustPlatform
+}:
+
+let
+  # Upstream only works with clang 16, so we're pinning it here to
+  # avoid future incompatibility.
+  # See https://docs.pgvecto.rs/developers/development.html#environment, step 4
+  rustPlatform' = rustPlatform // {
+    inherit (callPackage ../../../../../build-support/rust/hooks {
```

Comment:
Relative paths are not part of a package's interface. We'll have to find a different solution.

Reply:
You're absolutely right. I'll investigate alternative solutions

Reply:
There doesn't seem to be a way to override it. This is what I've tried:

```
nix-repl> pkgs.rustPlatform.bindgenHook.override { clang = pkgs.clang_17; }
error:
       … while calling a functor (an attribute set with a '__functor' attribute)

         at «string»:1:1:

            1| pkgs.rustPlatform.bindgenHook.override { clang = pkgs.clang_17; }
             | ^

       … while calling a functor (an attribute set with a '__functor' attribute)

         at /home/dtc/documents/vcs/nixpkgs/lib/customisation.nix:104:43:

          103|       # Re-call the function but with different arguments
          104|       overrideArgs = mirrorArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                           ^
          105|       # Change the result of the function call by applying g to it

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: function 'anonymous lambda' called with unexpected argument 'clang'

       at /home/dtc/documents/vcs/nixpkgs/pkgs/build-support/rust/hooks/default.nix:92:32:

           91|
           92|     bindgenHook = callPackage ({}: makeSetupHook {
             |                                ^
           93|       name = "rust-bindgen-hook";

```

```
nix-repl> (pkgs.makeRustPlatform {rustc = pkgs.rustc; cargo = pkgs.cargo; clang = pkgs.clang_17;}).bindgenHook.clang
«derivation /nix/store/j16048gsic45ijdpw82172pfxdz9nw2w-clang-wrapper-16.0.6.drv»
```

For now, I'll setup a similar script just for this package that supports pinning the clang version, but this should be revisited at a later date to allow overriding this.

Reply:
This looks like a bug. It can be fixed with this:

```patch
diff --git a/pkgs/development/compilers/rust/make-rust-platform.nix b/pkgs/development/compilers/rust/make-rust-platform.nix
index e22cb6d594af..6ed724aae821 100644
--- a/pkgs/development/compilers/rust/make-rust-platform.nix
+++ b/pkgs/development/compilers/rust/make-rust-platform.nix
@@ -1,4 +1,4 @@
-{ lib, buildPackages, callPackage, cargo-auditable, stdenv, runCommand }@prev:
+{ lib, buildPackages, callPackage, callPackages, cargo-auditable, stdenv, runCommand }@prev:
 
 { rustc
 , cargo
@@ -34,7 +34,7 @@ rec {
   };
 
   # Hooks
-  inherit (callPackage ../../../build-support/rust/hooks {
+  inherit (callPackages ../../../build-support/rust/hooks {
     inherit stdenv cargo rustc;
   }) cargoBuildHook cargoCheckHook cargoInstallHook cargoNextestHook cargoSetupHook maturinBuildHook bindgenHook;
 }
```

Reply:
I'll investigate that later then, thanks :)
Should this be made in another PR, or does it make sense to include it in this one?

Reply:
Great, it seems to work fine in the repl. I've tested it by overriding clang to 17 and having it fail the build.
I'm just finishing building the package to make sure everything works fine, and then I'll push.
This will likely trigger a lot of package rebuilds in the CI, but we'll see :upside_down_face:

Reply:
> Relative paths are not part of a package's interface. We'll have to find a different solution.

Sorry, I didn't understand that comment, and I'm curious about this... Why is the relative path a problem? Also, for example, `pkgs/os-specific/darwin/apple-sdk-11.0/default.nix` does something similar:
```
    rustPlatform = pkgs.makeRustPlatform {
      inherit (pkgs.darwin.apple_sdk_11_0) stdenv;
      inherit (pkgs) rustc cargo;
    } // {
      inherit (pkgs.callPackage ../../../build-support/rust/hooks {
        inherit (pkgs.darwin.apple_sdk_11_0) stdenv;
        inherit (pkgs) cargo rustc;
        clang = mkCc pkgs.clang;
      }) bindgenHook;
    };
```

Reply:
Relative paths are not guaranteed to remain stable.

People who reorganise one file in nixpkgs should not be expected to check and adjust relative paths in all other files.

This is relevant as we'll soon see the great RFC140 migration.

Now that the hooks override is fixed, the the apple SDK should use this style aswell as it doesn't have this obvious issue. (It may still have less obvious issues.)